src/timely-util/src/probe.rs

```diff
@@ -0,0 +1,146 @@
+use timely::progress::Timestamp;
+use timely::Data;
+use tokio::sync::Notify;
+
+/// Monitors progress at a `Stream`.
+trait ProbeNotify<G: Scope, D: Data> {
+    /// Constructs a progress probe which indicates which timestamps have elapsed at the operator.
+    fn probe_notify(&self) -> Handle<G::Timestamp>;
+
+    /// Inserts a progress probe in a stream.
+    fn probe_notify_with(&self, handle: &mut Handle<G::Timestamp>) -> Stream<G, D>;
+}
+
+impl<G: Scope, D: Data> ProbeNotify<G, D> for Stream<G, D> {
+    fn probe_notify(&self) -> Handle<G::Timestamp> {
+        let mut handle = Handle::default();
+        self.probe_notify_with(&mut handle);
+        handle
+    }
+
+    fn probe_notify_with(&self, handle: &mut Handle<G::Timestamp>) -> Stream<G, D> {
+        let mut handle = handle.clone();
+        self.inspect_container(move |update| {
+            if let Err(frontier) = update {
+                handle.downgrade(frontier);
+            }
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Handle<T: Timestamp> {
+    /// The overall shared frontier managed by all the handles
+    frontier: Rc<RefCell<MutableAntichain<T>>>,
+    /// The private frontier containing the changes produced by this handle only
+    handle_frontier: Antichain<T>,
+    notify: Rc<Notify>,
+}
+
+impl<T: Timestamp> Default for Handle<T> {
+    fn default() -> Self {
+        Handle {
+            frontier: Rc::new(RefCell::new(MutableAntichain::new_bottom(T::minimum()))),
+            handle_frontier: Antichain::from_elem(T::minimum()),
+            notify: Rc::new(Notify::new()),
+        }
+    }
+}
+
+impl<T: Timestamp> Handle<T> {
+    /// Wait for the frontier monitored by this probe to progress
+    pub async fn progressed(&self) {
+        self.notify.notified().await
+    }
+
+    /// Returns true iff the frontier is strictly less than `time`.
+    #[inline]
+    pub fn less_than(&self, time: &T) -> bool {
+        self.frontier.borrow().less_than(time)
+    }
+    /// Returns true iff the frontier is less than or equal to `time`.
+    #[inline]
+    pub fn less_equal(&self, time: &T) -> bool {
+        self.frontier.borrow().less_equal(time)
+    }
+    /// Returns true iff the frontier is empty.
+    #[inline]
+    pub fn done(&self) -> bool {
+        self.frontier.borrow().is_empty()
+    }
+
+    /// Invokes a method on the frontier, returning its result.
+    ///
+    /// This method allows inspection of the frontier, which cannot be returned by reference as
+    /// it is on the other side of a `RefCell`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use mz_timely_util::probe::Handle;
+    ///
+    /// let handle = Handle::<usize>::default();
+    /// let frontier = handle.with_frontier(|frontier| frontier.to_vec());
+    /// ```
+    #[inline]
+    pub fn with_frontier<R, F: FnMut(AntichainRef<T>) -> R>(&self, mut function: F) -> R {
+        function(self.frontier.borrow().frontier())
+    }
+
+    #[inline]
```

Comment:
Curious about when it makes sense to add explicit `inline` annotations. Is there some general guidance for this?

Reply:
No idea, I shamefully copied them from timely where I suppose were added after some thought